### PR TITLE
deprecate auto-idna on UniformResourceIdentifier

### PR DESCRIPTION
**Pull request by reaperhulk**
refs #3357

### CODE REVIEW

File: src/cryptography/x509/general_name.py [hunk 1]
```diff
@@ -173,13 +173,34 @@ def __ne__(self, other):
...
-        if not isinstance(value, six.text_type):
-            raise TypeError("value must be a unicode string")
+        if isinstance(value, six.text_type):
+            try:
+                value = value.encode("ascii")
+            except UnicodeEncodeError:
+                value = self._idna_encode(value)
+                warnings.warn(
+                    "UniformResourceIdentifier values should be passed as "
+                    "idna-encoded bytes, not strings. Support for passing "
```
Comment (alex) [inline src/cryptography/x509/general_name.py]: `idna-encoded` is an over simplification right? the hostname portion should be idna-encoded, but the path segment should use the urlquote stuff, right?

File: src/cryptography/x509/general_name.py [hunk 2]
```diff
@@ -173,13 +173,34 @@ def __ne__(self, other):
...
+                    "unicode strings will be removed in a future version.",
+                    utils.DeprecatedIn21,
+                    stacklevel=2,
+                )
+            else:
+                warnings.warn(
+                    "UniformResourceIdentifier values should be passed as "
+                    "idna-encoded bytes, not strings. Support for passing "
```
Comment (alex) [inline src/cryptography/x509/general_name.py]: same

File: src/cryptography/x509/general_name.py (hunk 1 above)
Comment (reaperhulk) [inline src/cryptography/x509/general_name.py]: Yeah, maybe we just document it as "URIs should be passed as bytes with the hostname idna encoded"?

Comment (alex) [inline src/cryptography/x509/general_name.py]: Sounds right.

**Pull request merged**